IndexOutOfBoundsException in content assist / FollowElementComputer

After programming N4JS (in Eclipse) for a while I had several of the following stack trace in the Console view of the parent Eclipse instance:
```
2020-05-127-01:46:31.031+0200 -                      main - ERROR -    ContentAssistContextFactory - (org.eclipse.xtext.ide.editor.contentassist.antlr.ContentAssistContextFactory.java:182) - java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
java.util.concurrent.ExecutionException: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.ContentAssistContextFactory.doCreateContexts(ContentAssistContextFactory.java:180)
	at org.eclipse.n4js.ide.editor.contentassist.N4JSContentAssistContextFactory.doCreateContexts(N4JSContentAssistContextFactory.java:40)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.ContentAssistContextFactory.create(ContentAssistContextFactory.java:130)
	at org.eclipse.xtext.ui.editor.contentassist.antlr.DelegatingContentAssistContextFactory$StatefulFactory.doCreateContexts(DelegatingContentAssistContextFactory.java:92)
	at org.eclipse.xtext.ui.editor.contentassist.antlr.DelegatingContentAssistContextFactory$StatefulFactory.create(DelegatingContentAssistContextFactory.java:78)
	at org.eclipse.xtext.ui.editor.contentassist.antlr.DelegatingContentAssistContextFactory.create(DelegatingContentAssistContextFactory.java:55)
	at org.eclipse.n4js.ui.utils.ImportUtil.createContentAssistContexts(ImportUtil.java:158)
	at org.eclipse.n4js.ui.utils.ImportUtil.lambda$0(ImportUtil.java:145)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:91)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:524)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.readOnly(XtextDocument.java:496)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.readOnly(XtextDocument.java:135)
	at org.eclipse.n4js.ui.utils.ImportUtil.findImportCandidates(ImportUtil.java:153)
	at org.eclipse.n4js.ui.utils.ImportUtil.findImportCandidates(ImportUtil.java:129)
	at org.eclipse.n4js.ui.quickfix.N4JSQuickfixProvider.addImportForUnresolvedReference(N4JSQuickfixProvider.java:303)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.xtext.ui.editor.quickfix.AbstractDeclarativeQuickfixProvider.getResolutions(AbstractDeclarativeQuickfixProvider.java:57)
	at org.eclipse.xtext.ui.editor.quickfix.AbstractDeclarativeQuickfixProvider.getResolutions(AbstractDeclarativeQuickfixProvider.java:87)
	at org.eclipse.xtext.ui.editor.quickfix.DefaultQuickfixProvider.getResolutions(DefaultQuickfixProvider.java:246)
	at org.eclipse.xtext.ui.editor.quickfix.AbstractIssueResolutionProviderAdapter.getResolutions(AbstractIssueResolutionProviderAdapter.java:46)
	at org.eclipse.xtext.ui.editor.quickfix.XtextQuickAssistProcessor.createQuickfixes(XtextQuickAssistProcessor.java:154)
	at org.eclipse.xtext.ui.editor.quickfix.XtextQuickAssistProcessor.computeQuickAssistProposals(XtextQuickAssistProcessor.java:123)
	at org.eclipse.xtext.ui.editor.hover.AnnotationWithQuickFixesHover$CompletionProposalRunnable.run(AnnotationWithQuickFixesHover.java:604)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:185)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4145)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3812)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1049)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:633)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:557)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
	at org.eclipse.n4js.product.N4JSApplication.start(N4JSApplication.java:141)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:400)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:660)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:597)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1468)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1441)
Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:372)
	at java.base/java.util.ArrayList.get(ArrayList.java:458)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.FollowElementComputer.setParamConfigAndUpdateOffset(FollowElementComputer.java:174)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.FollowElementComputer.computeFollowElements(FollowElementComputer.java:99)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.FollowElementComputer.computeFollowElements(FollowElementComputer.java:164)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.FollowElementComputer.computeFollowElements(FollowElementComputer.java:85)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.FollowElementComputer.computeFollowElements(FollowElementComputer.java:75)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.FollowElementComputer.computeFollowElements(FollowElementComputer.java:80)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.ContentAssistContextFactory.doCreateContexts(ContentAssistContextFactory.java:318)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.PartialContentAssistContextFactory.handleLastCompleteNodeAsPartOfDatatypeNode(PartialContentAssistContextFactory.java:53)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.ContentAssistContextFactory$1.call(ContentAssistContextFactory.java:148)
	at org.eclipse.xtext.ide.editor.contentassist.antlr.ContentAssistContextFactory$1.call(ContentAssistContextFactory.java:145)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

--- a/plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/editor/contentassist/CustomN4JSParser.java
+++ b/plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/editor/contentassist/CustomN4JSParser.java
@@ -71,15 +71,22 @@ import com.google.inject.Inject;
 public class CustomN4JSParser extends N4JSParser implements IPartialContentAssistParser {
 
 	/**
-	 * Overridden to make protected member accessible from within the {@link CustomN4JSParser}.
+	 * Overridden to make protected member accessible from within the {@link CustomN4JSParser} and to fix GH-1756
 	 */
 	protected static class CustomInternalN4JSParser extends InternalN4JSParser {
+
+		/**
+		 * The parameter stack as it was seen at a given look-ahead position. Cleared whenever the parser consumes a
+		 * token. The same apporach is used for the inherited {@link #getLocalTrace()}.
+		 */
+		protected final List<Integer> localParamStack;
 
 		/**
 		 * Default constructor.
 		 */
 		public CustomInternalN4JSParser() {
 			super(null);
+			this.localParamStack = new ArrayList<>();
 		}
 
 		/**
@@ -87,6 +94,39 @@ public class CustomN4JSParser extends N4JSParser implements IPartialContentAssis
 		 */
 		public RecognizerSharedState getState() {
 			return state;
+		}
+
+		@Override
+		public void announceConsume() {
+			if (marked <= 0) {
+				localParamStack.clear();
+			}
+			super.announceConsume();
+		}
+
+		/*
+		 * This is here to provoke a compile error against a version of Xtext that has an equivalent fix for GH-1756.
+		 */
+		private List<Integer> getLocalParamStack() {
+			return localParamStack;
+		}
+
+		@Override
+		public void before(EObject grammarElement, int paramConfig) {
+			super.before(grammarElement, paramConfig);
+			localParamStack.add(paramConfig);
+		}
+
+		@Override
+		protected DefaultFollowElementFactory newFollowElementFactory() {
+			return new DefaultFollowElementFactory() {
+				@Override
+				public FollowElement createFollowElement(AbstractElement current, int lookAhead) {
+					FollowElement result = super.createFollowElement(current, lookAhead);
+					result.getParamStack().addAll(getLocalParamStack());
+					return result;
+				}
+			};
 		}
 
 	}

--- a/tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/contentassist/Issue1756Test.xtend
+++ b/tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/contentassist/Issue1756Test.xtend
@@ -1,0 +1,65 @@
+/** 
+ * Copyright (c) 2020 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ * Contributors:
+ * NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.ide.tests.contentassist
+
+import org.eclipse.n4js.ide.tests.server.AbstractCompletionTest
+import org.junit.Test
+
+/**
+ * see https://github.com/eclipse/n4js/issues/1756
+ */
+class Issue1756Test extends AbstractCompletionTest {
+
+	@Test
+	def void test01() {
+		testAtCursor('''
+			let value;
+			
+			export function foo(): any {
+				return (arg) => {
+					let x = async () => {
+						value = new C((p: Object) => {
+							someFun<|>
+						});
+						return null;
+					};
+				}
+			}
+			
+			class C {
+				constructor(f: Function) {}
+			}
+			
+			function someFunctionWithAnExtremelyLooooongName() {}
+		''', ''' 
+			(someFunctionWithAnExtremelyLooooongName, Function, someFunctionWithAnExtremelyLooooongName, , , 00000, , , , ([6:4 - 6:11], someFunctionWithAnExtremelyLooooongName), [], [], , )
+		''');
+	}
+
+	@Test
+	def void test02() {
+		testAtCursor('''
+			export function xxxxx(): any {
+				((arg) => {
+					let x = async () => {
+						value = new C((p: Object) => {
+							xxx<|>
+						});
+					};
+				})
+			}
+			function xxxx() {}
+		''', ''' 
+			(xxxx, Function, xxxx, , , 00000, , , , ([4:4 - 4:7], xxxx), [], [], , )
+			(xxxxx, Function, MyModule, , , 00001, , , , ([4:4 - 4:7], xxxxx), [], [], , )
+		''');
+	}
+
+}

--- a/tests/org.eclipse.n4js.ui.tests/src/org/eclipse/n4js/tests/contentAssist/Issue1756PluginUITest.xtend
+++ b/tests/org.eclipse.n4js.ui.tests/src/org/eclipse/n4js/tests/contentAssist/Issue1756PluginUITest.xtend
@@ -1,0 +1,44 @@
+/**
+ * Copyright (c) 2020 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.tests.contentAssist
+
+import org.junit.Test
+
+/**
+ *
+ */
+class Issue1756PluginUITest extends AbstractN4JSContentAssistPluginUITest {
+
+	@Test def void test_01() throws Exception {
+		newBuilder().append('''
+			let value;
+			
+			export function foo(): any {
+			    return (arg) => {
+			        let x = async () => {
+			            value = new C((p: Object) => {
+			            	someFun<|>
+			            });
+			            return null;
+			        };
+				}
+			}
+			
+			class C {
+				constructor(f: Function) {}
+			}
+			
+			function someFunctionWithAnExtremelyLooooongName() {}
+		'''
+		).assertProposalAtCursor('someFunctionWithAnExtremelyLooooongName');
+	}
+	
+}

